src/aiortc/contrib/media.py

```diff
@@ -155,21 +165,40 @@ def player_worker(
             if video_first_pts is None:
                 video_first_pts = frame.pts
             frame.pts -= video_first_pts
-            print("Video frame %s retrieved: %s" % (str(frame.index), str(frame)))
+            logger.warning(
+                "MediaPlayer(%s) Video frame %s retrieved: %s", container.name, str(frame.index), str(frame)
+            )
             frame_time = frame.time
             asyncio.run_coroutine_threadsafe(video_track._queue.put(frame), loop)
 
             # Extract the keypoints from the frame
-            keypoints_generator = KeypointsGenerator()
-            try:
-                keypoints = keypoints_generator.get_keypoints(frame.to_rgb().to_ndarray())
-                keypoints_frame = KeypointsFrame(keypoints, frame.pts)
-                print("Keypoints for frame index %s retrieved." % str(frame.index))
-            except:
-                keypoints_frame = KeypointsFrame(bytes("Error!", encoding='utf8'), frame.pts)
-                print("Could not extract the keypoints for frame index %s" % str(frame.index))
+            if enable_prediction:
+                try:
+                    frame_array = frame.to_rgb().to_ndarray()
+                    time_before_keypoints = time.time()
+                    keypoints =  model.extract_keypoints(frame_array)
+                    time_after_keypoints = time.time()
+                    logger.warning(
+                        "MediaPlayer(%s) Keypoints extract time for frame index %s in sender: %s",
+                        container.name, str(frame.index), str(time_after_keypoints - time_before_keypoints)
+                    )
+                    keypoints_frame = KeypointsFrame(keypoints, frame.pts) 
+                    
+                    if frame.index % UPDATE_SRC_FREQ == 0:
+                        time_before_update = time.time()
+                        model.update_source(frame_array, keypoints)
+                        time_after_update = time.time()
+                        logger.warning(
+                            "MediaPlayer(%s) Source update time for frame index %s in sender: %s",
+                            container.name, str(frame.index), str(time_after_update - time_before_update)
+                        )
+                except:
+                    keypoints_frame = KeypointsFrame(bytes("Error!", encoding='utf8'), frame.pts)
+                    logger.warning(
+                        "MediaPlayer(%s) Could not extract the keypoints for frame index %s", str(frame.index)
+                    )
 
-            asyncio.run_coroutine_threadsafe(keypoints_track._queue.put(keypoints_frame), loop)
+                asyncio.run_coroutine_threadsafe(keypoints_track._queue.put(keypoints_frame), loop)
```

Comment:
will this cause issue if the keypoints frame is an "error"? like on the receive prediction side? Is there a better way to handle error?

Reply:
Yes, it will cause an issue. Task 4 : `on bug_fix branch: Keypoint extraction failing`
If we don't have a keypoint, we should send a source frame, but we need to make a mechanism to signal the receiver to show the original video frame